exonum/src/crypto.rs

```diff
@@ -529,6 +542,97 @@ impl Default for Hash {
     }
 }
 
+impl CryptoHash for u8 {
+    fn hash(&self) -> Hash {
+        hash(&[*self])
+    }
+}
+
+impl CryptoHash for u16 {
+    fn hash(&self) -> Hash {
+        let mut v = [0; 2];
+        LittleEndian::write_u16(&mut v, *self);
+        hash(&v)
+    }
+}
+
+impl CryptoHash for u32 {
+    fn hash(&self) -> Hash {
+        let mut v = [0; 4];
+        LittleEndian::write_u32(&mut v, *self);
+        hash(&v)
+    }
+}
+
+impl CryptoHash for u64 {
+    fn hash(&self) -> Hash {
+        let mut v = [0; 8];
+        LittleEndian::write_u64(&mut v, *self);
+        hash(&v)
+    }
+}
+
+impl CryptoHash for i8 {
+    fn hash(&self) -> Hash {
+        hash(&[*self as u8])
+    }
+}
+
+impl CryptoHash for i16 {
+    fn hash(&self) -> Hash {
+        let mut v = [0; 2];
+        LittleEndian::write_i16(&mut v, *self);
+        hash(&v)
+    }
+}
+
+impl CryptoHash for i32 {
+    fn hash(&self) -> Hash {
+        let mut v = [0; 4];
+        LittleEndian::write_i32(&mut v, *self);
+        hash(&v)
+    }
+}
+
+impl CryptoHash for i64 {
+    fn hash(&self) -> Hash {
+        let mut v = [0; 8];
+        LittleEndian::write_i64(&mut v, *self);
+        hash(&v)
+    }
+}
+
+impl CryptoHash for () {
```

Comment:
I cannot help but notice that `()` and `Hash` implementations of the trait are somewhat inconsistent: a SHA-256 hash of the empty byte array does not equal `[0; 32]`, and it's not obvious why `Hash` values shouldn't be hashed.

Reply:
Hm, fascinating! 🤔 

I would think that hashes should be hashed: one of the requirements of the cryptographic hash functions is that you can't reconstruct the original value significantly more efficient than by just enumerating all possible values, and identity function certainly does not have this property. 

For `()`, hashing an empty array for consistency with `StorageValue` also makes sense. Basically, hashing should work as if there were a blanket `impl<T: StorageValue> CryptoHash for T { ... }`.

Should we make both changes? I would say yes, but his would be a storage breaking change :( 

Should we make these changes in this PR? I'd say no :)

Reply:
Storage breaking is OK for me.

Separate PR is a must. Would you like to do that? There are other people if not. :smile:

Reply:
Hm, @vldm pointed out that `Message::hash` and `StorageValue::hash` have slightly different usages.

`StorageValue::hash` is primarily used in indices like `ValueSetIndex`, and for this use-case you probably don't need strong cryptography guarantees, besides "effectively unique" hash. That is, it is OK to use identify function, and it is even probably ok to *not* hash integers and gain some performance out of it.

`Message::hash` on the other hand is used to identify transactions, and must actually be a hash of raw bytes.

However, using the message-style hash for both use-cases seems valid (and this is done for integers already), so I suggest that we do use approach in this PR, and do change `CryptoHash` for `()` and `Hash` in a follow up PR.